internal/service/creditgrant.go

```diff
@@ -219,3 +215,154 @@ func (s *creditGrantService) GetCreditGrantsBySubscription(ctx context.Context,
 
 	return resp, nil
 }
+
+// ProcessRecurringGrants processes all recurring credit grants
+func (s *creditGrantService) ProcessRecurringGrants(ctx context.Context) error {
+	s.Logger.Infow("starting recurring credit grant processing")
+
+	// 1. Find all active recurring grants
+	grants, err := s.CreditGrantRepo.FindActiveRecurringGrants(ctx)
+	if err != nil {
+		return err
+	}
+
+	s.Logger.Infow("found active recurring grants", "count", len(grants))
+
+	// 2. Process each grant
+	for _, grant := range grants {
+		if err := s.processRecurringGrant(ctx, grant); err != nil {
+			s.Logger.Errorw("failed to process recurring grant",
+				"grant_id", grant.ID,
+				"error", err)
+			// Continue processing other grants
+		}
+	}
+
+	return nil
+}
+
+func (s *creditGrantService) processRecurringGrant(ctx context.Context, grant *creditgrant.CreditGrant) error {
+	// Find eligible subscriptions for this grant
+	subscriptions, err := s.SubRepo.FindEligibleSubscriptionsForGrant(ctx, grant)
+	if err != nil {
+		return err
+	}
+
+	s.Logger.Debugw("found eligible subscriptions for grant",
+		"grant_id", grant.ID,
+		"subscription_count", len(subscriptions))
+
+	for _, subscription := range subscriptions {
+		if err := s.ProcessGrantForSubscription(ctx, grant, subscription); err != nil {
+			s.Logger.Errorw("failed to process grant for subscription",
+				"grant_id", grant.ID,
+				"subscription_id", subscription.ID,
+				"error", err)
+			// Continue with other subscriptions
+		}
+	}
+
+	return nil
+}
+
+func (s *creditGrantService) ProcessGrantForSubscription(ctx context.Context, grant *creditgrant.CreditGrant, subscription *subscription.Subscription) error {
+	// Determine if grant should be applied based on subscription state
+	stateHandler := NewSubscriptionStateHandler(subscription, grant)
+
+	action, reason := stateHandler.DetermineAction()
+
+	// Apply the determined action
+	switch action {
+	case StateActionApply:
+		return s.applyRecurringGrant(ctx, grant, subscription)
+	case StateActionSkip:
+		s.Logger.Debugw("skipping grant application", "reason", reason, "grant_id", grant.ID, "subscription_id", subscription.ID)
+		return nil
+	case StateActionDefer:
+		s.Logger.Debugw("deferring grant application", "reason", reason, "grant_id", grant.ID, "subscription_id", subscription.ID)
+		return nil
+	case StateActionCancel:
+		s.Logger.Debugw("cancelling grant application", "reason", reason, "grant_id", grant.ID, "subscription_id", subscription.ID)
+		return nil
+	default:
+		return ierr.NewError("unknown state action").Mark(ierr.ErrInternal)
+	}
+}
+
+func (s *creditGrantService) applyRecurringGrant(ctx context.Context, grant *creditgrant.CreditGrant, subscription *subscription.Subscription) error {
```

Comment:
can we merge/or reuse 90% of code between this and the handleCreditGrant function in subscription service so that we do not have duplicate logic

this function can be made aware of recurring/one time config as simple as that